crates/telio-firewall/src/firewall.rs

```diff
@@ -3137,4 +3142,158 @@ pub mod tests {
+            let tcp_ack_inbound = make_tcp(dst1, src1, TcpFlags::ACK);
+
+            let udp_outbound = make_udp(src1, dst1);
+            let udp_inbound = make_udp(dst1, src1);
+
+            let mut data = vec![1, 0, 0, 1];
+            let tcp = make_tcp(src1, dst1, TcpFlags::ACK);
+            data.append(&mut tcp.clone());
+            let icmp_tcp_error_inbound =
+                make_icmp4_with_body(dst2, src2, IcmpTypes::DestinationUnreachable.into(), &data);
+
+            // tcp & udp outbound: allowed
+            assert!(fw.process_outbound_packet(&good_peer.0, &tcp_syn_outbound),);
+            // inject error package from bad peer
+            assert!(!fw.process_inbound_packet(&bad_peer.0, &icmp_tcp_error_inbound),);
+            // expect that connection continues normally
+            assert!(fw.process_inbound_packet(&good_peer.0, &tcp_ack_inbound),);
+        }
+    }
+
+    #[test]
+    fn firewall_tcp_rst_attack_vulnerability() {
+        struct TestInput {
+            us: &'static str,
+            them: &'static str,
+            random: &'static str,
+        }
+
+        let test_inputs = [TestInput {
+            us: "127.0.0.1:1111",
+            them: "8.8.8.8:8888",
+            random: "192.168.0.1:7777",
+        }];
+        for test_input @ TestInput { us, them, random } in test_inputs {
+            let fw = StatefullFirewall::new_custom(3, LRU_TIMEOUT, false, false);
+
+            let outgoing_init_packet = make_tcp(us, them, TcpFlags::SYN);
+            let incoming_rst_packet = make_tcp(them, us, TcpFlags::RST);
+
+            let peer_bad = make_random_peer();
+            let peer_good = make_peer();
+
+            assert!(fw.process_outbound_packet(&peer_good, &outgoing_init_packet),);
+            assert!(!fw.process_inbound_packet(&peer_bad.0, &incoming_rst_packet),);
+            assert!(fw.process_inbound_packet(
+                &peer_good,
+                &make_tcp(them, us, TcpFlags::SYN | TcpFlags::ACK)
+            ),);
+            assert!(fw.process_inbound_packet(&peer_good, &make_tcp(them, us, TcpFlags::SYN)),);
+        }
+    }
+
+    #[test]
+    fn firewall_udp_spoffing_and_injection_vulnerability() {
+        struct TestInput {
+            src1: &'static str,
+            src2: &'static str,
+            src3: &'static str,
+            src4: &'static str,
+            src5: &'static str,
+            dst1: &'static str,
+            dst2: &'static str,
+            make_udp: MakeUdp,
+        }
+
+        let test_inputs = [
+            TestInput {
+                src1: "127.0.0.1:1111",
+                src2: "127.0.0.1:2222",
+                src3: "127.0.0.1:3333",
+                src4: "127.0.0.1:4444",
+                src5: "127.0.0.1:5555",
+                dst1: "8.8.8.8:8888",
+                dst2: "8.8.8.8:7777",
+                make_udp: &make_udp,
+            },
+            // TODO(msz): Implement a fix for IPv6
```

Comment:
If there is a separate ticket for fixing this then I think you should add the ticket number here, otherwise, I think it should be fixed in this PR

Reply:
I've cleaned the test code. However since each test case corresponds to a different VCP ticket, I prefer to keep them in separate commits